reference/src/representation/function-pointers.md

```diff
@@ -0,0 +1,116 @@
+# Representation of Function Pointers
+
+### Terminology
+
+In Rust, a function pointer type, is either `fn(Args...) -> Ret`,
+`extern "ABI" fn(Args...) -> Ret`, `unsafe fn(Args...) -> Ret`, and
+`unsafe extern "ABI" fn(Args...) -> Ret`.
+A function pointer is the address of a function,
+and has function pointer type.
+The pointer is implicit in the `fn` type,
+and `fn` types are implicitly `'static`.
+
+In C, a function pointer type is `Ret (*)(Args...)`, or `Ret ABI (*)(Args...)`,
+and values of function pointer type are either a null pointer value,
+or the address of a function.
+
+### Representation
+
+The ABI and layout of `(unsafe)? (extern "ABI")? fn(Args...) -> Ret`
+is exactly that of the corresonding C type --
+the lack of a null value does not change this. 
+On common platforms, this means that `*const ()` and `fn(Args...) -> Ret` have
+the same ABI and layout. This is, in fact, guaranteed by POSIX and Windows.
+This means that for the vast majority of platforms,
+
+```rust
+fn go_through_pointer(x: fn()) -> fn() {
+  let ptr = x as *const ();
+  unsafe { std::mem::transmute::<*const (), fn()>(ptr) }
+}
+```
+
+is both perfectly safe, and, in fact, required for some APIs -- notably,
+`GetProcAddress` on Windows requires you to convert from `void (*)()` to
+`void*`, to get the address of a variable;
+and the opposite is true of `dlsym`, which requires you to convert from
+`void*` to `void (*)()` in order to get the address of functions.
+This conversion is _not_ guaranteed by Rust itself, however;
+simply the implementation. If the underlying platform allows this conversion,
+so will Rust.
+
+However, null values are not supported by the Rust function pointer types --
+just like references, the expectation is that you use `Option` to create
+nullable pointers. `Option<fn(Args...) -> Ret>` will have the exact same ABI
+as `fn(Args...) -> Ret`, but additionally allows null pointer values.
```

Comment:
This should likely reference the `enum` chapter once one exists; I guess the rules here will be exactly the same as the ones for reference types?

Reply:
Yep!